transformer_nuggets/llama/finetune.py

```diff
@@ -0,0 +1,356 @@
+    world_size: int,
+):
+    torch.cuda.set_device(rank)
+
+    random.seed(1337)
+    np.random.seed(1337)
+    torch.manual_seed(1337)
+    torch.cuda.manual_seed_all(1337)
+
+    os.makedirs(training_config.out_dir, exist_ok=True)
+    os.makedirs(training_config.log_dir, exist_ok=True)
+
+    # Setup Model
+    model_args = ModelArgs.from_name(training_config.model_name)
+    if rank == 0:
+        logging.info(f"Initializing model: {training_config.model_name}")
+    with training_config.device:
+        model = Transformer(model_args).to(torch.bfloat16)
+        model.init_parameters()
+
+        qlora_config = qlora.QloraConfig(
+            hyper_params.lora_r,
+            hyper_params.lora_alpha,
+            hyper_params.lora_dropout,
+        )
+        qlora.swap_for_qlora(model, qlora_config, torch.bfloat16)
+    model.setup_caches(
+        hyper_params.micro_batch_size, hyper_params.max_seq_length, training_config.device
+    )
+
+    if rank == 0:
+        logging.info("Setting up the dataloaders")
+    train_data, val_data = load_datasets(hyper_params, training_config, rank, world_size)
+    train_dataloader = DataLoader(
+        train_data,
+        batch_size=hyper_params.micro_batch_size,
+        num_workers=2,
+    )
+    val_dataloader = DataLoader(val_data, batch_size=hyper_params.micro_batch_size, num_workers=2)
+
+    log_num_params(model)
+
+    if world_size > 1:
+        model = FSDP(
+            model,
+            use_orig_params=True,
+            auto_wrap_policy=ModuleWrapPolicy([TransformerBlock]),
+        )
+
+    if training_config.compile:
+        model = torch.compile(model)
+
+    if rank == 0:
+        print(model)
+
+    optimizer = torch.optim.AdamW(
+        [p for p in model.parameters() if p.requires_grad],
+        lr=hyper_params.learning_rate,
+        weight_decay=hyper_params.weight_decay,
+        betas=(hyper_params.beta1, hyper_params.beta2),
+        foreach=hyper_params.foreach_optimizer,
+    )
+
+    train(
+        model,
+        optimizer,
+        train_dataloader,
+        val_dataloader,
+        hyper_params,
+        training_config,
+        rank,
+        world_size,
+    )
+
+
+def entrypoint(
+    profile: bool = False,
+    rank: int = 0,
+    world_size: int = 1,
+):
+    batch_size = int(128 / world_size)
+    assert isinstance(profile, bool), "profile must be bool"
+    hyper_params = Hyperparameters(batch_size=batch_size)
+    training_config = TrainingConfig(
+        profile=profile,
+        device=torch.device(f"cuda:{rank}"),
+    )
+    main(hyper_params, training_config, rank, world_size)
+
+
+def fsdp_main(rank, world_size, args):
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+    dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    entrypoint(*args, rank=rank, world_size=world_size)
+    dist.destroy_process_group()
+
+
+# copied from https://github.com/pytorch-labs/ao_benchmarks/tree/main/llama
```

Comment:
lets remove this

Reply:
remove the comment itself or avoid copying code?

Reply:
I removed the comment